### PR TITLE
Add icu_properties feature to icu_normalizer

### LINKED ISSUE #5121: Move CanonicalCombiningClass type out of icu_properties
The data for [CanonicalCombiningClass](https://unicode-org.github.io/icu4x/rustdoc/icu_properties/struct.CanonicalCombiningClass.html#) is available both via icu_properties and via icu_normalizer. To have the same type in both cases, `icu_normalizer` depends on `icu_properties`. This prevents the two from being compiled in parallel. See https://github.com/servo/rust-url/issues/939

We should break the dependency. The obvious thing to do is to move the type to `icu_normalizer` and not make the data available via `icu_properties` at all. If we really wanted to have two ways of providing the data, the type should go into a third crate, but there isn't a candidate destination that wouldn't be weird.

### PR DESCRIPTION
**Pull request by Manishearth**
Fixes #5121

### CODE REVIEW

File: components/normalizer/src/lib.rs [hunk 1]
```diff
@@ -67,6 +67,18 @@
 
 extern crate alloc;
 
+// We don't depend on icu_properties to minimize deps, but we want to be able
+// to ensure we're using the right CCC values
+macro_rules! ccc {
+    ($name:ident, $num:expr) => {{
+        #[cfg(all(debug_assertions, feature = "icu_properties"))]
+        if icu_properties::CanonicalCombiningClass::$name.0 != $num {
+            panic!("icu_normalizer has incorrect ccc values")
```
Comment (sffc) [inline components/normalizer/src/lib.rs:76]: Praise: A static assertion?! Since when has that worked?

**Review by sffc — APPROVED** [review]

**Comment by sffc** [comment]
```
error[E0599]: no method named `get32` found for struct `CanonicalCombiningClassMapBorrowed` in the current scope
  --> ffi/capi/src/normalizer_properties.rs:54:34
   |
54 |             self.0.as_borrowed().get32(ch).0
   |                                  ^^^^^ help: there is a method with a similar name: `get32_u8`

For more information about this error, try `rustc --explain E0599`.
error: could not compile `icu_capi` (lib) due to previous error

```

**Comment by hsivonen** [comment]
Thanks. This needs testing with Pernosco to see if this break Pernosco's special knowledge about the ICU4X types involved here. I'll test and will report back.

**Comment by hsivonen** [comment]
I created a new crate that depends on `icu_normalizer` by `path`, and building the new crate ends up building `icu_properties` anyway. Is this changeset accomplishing what's intended?

**Review by hsivonen — DISMISSED** [review]
This changeset with or without the feature enabled breaks Pernosco's rendering of the `CharacterAndClass` type.

Without this PR:
self ::CharacterAndClass*@0x7285c15ff1b8={char=‘⁨̧⁩’ 0x327, ccc=(202 ‘Ê’)}

With this PR:
self ::CharacterAndClass*@0x7eacb49ff1b8=(3388998439)

I think this should not be merged as-is.

I'm not quite sure what precisely Pernosco wants to see. I suggest putting the accessors on `CharacterAndClass` that have `CanonicalCombiningClass` in the return value behind the feature and introducing differently-named `u8` accessors. (I haven't yet tested if that would work.)

**Comment by hsivonen** [comment]
> I created a new crate that depends on `icu_normalizer` by `path`, and building the new crate ends up building `icu_properties` anyway. Is this changeset accomplishing what's intended?

I pushed a change to restore Pernosco compat. (That too _way_ more rounds of experimentation than I expected!) However, the above observation still applies: `icu_properties` gets compiled when the `icu_properties` feature isn't set. Why?

File: components/normalizer/Cargo.toml [hunk 2]
```diff
@@ -22,7 +22,7 @@ all-features = true
 [dependencies]
 displaydoc = { workspace = true }
 icu_collections = { workspace = true }
-icu_properties = { workspace = true }
+icu_properties = { workspace = true, optional = true, default-features = false }
```
Comment (robertbastian) [inline components/normalizer/Cargo.toml:25]: workspace already sets default-features to false

File: components/normalizer/src/lib.rs (hunk 1 above)
Comment (robertbastian) [reply components/normalizer/src/lib.rs:76]: we've been using panicking in consts for years, that's how all the `icu::locale` macros work.

File: components/normalizer/src/lib.rs [hunk 3]
```diff
@@ -180,12 +196,12 @@ fn decomposition_starts_with_non_starter(trie_value: u32) -> bool {
 ///
 /// The trie value must not be one that signifies a special non-starter
 /// decomposition. (Debug-only)
-fn ccc_from_trie_value(trie_value: u32) -> CanonicalCombiningClass {
+fn ccc_from_trie_value(trie_value: u32) -> u8 {
```
Comment (robertbastian) [inline components/normalizer/src/lib.rs:199]: nit: define `type CanonicalCombiningClass = u8` or a newtype, and then qualify the properties one?

File: components/normalizer/src/lib.rs [hunk 4]
```diff
@@ -67,6 +67,18 @@
 
 extern crate alloc;
 
+// We don't depend on icu_properties to minimize deps, but we want to be able
+// to ensure we're using the right CCC values
+macro_rules! ccc {
+    ($name:ident, $num:expr) => {{
+        #[cfg(all(debug_assertions, feature = "icu_properties"))]
```
Comment (robertbastian) [inline components/normalizer/src/lib.rs:74]: wouldn't even gate this on debug_assertions

Comment (Manishearth) [reply components/normalizer/src/lib.rs:74]: I'm using it in non const contexts too. I could make consts for everything

Comment (robertbastian) [reply components/normalizer/src/lib.rs:74]: You don't need to define consts, the macro can produce a const expression:

```rust
{ 
  const X: u8 = {
    #[cfg(all(debug_assertions, feature = "icu_properties"))]
    if icu_properties::CanonicalCombiningClass::$name.0 != $num {
      panic!("icu_normalizer has incorrect ccc values")
    }
    $num
  };
  X
}
```

**Comment by Manishearth** [comment]
> However, the above observation still applies: `icu_properties` gets compiled when the `icu_properties` feature isn't set. Why?


It's because I didn't switch to `?` deps. I fixed it.

File: components/normalizer/src/lib.rs (hunk 3 above)
Comment (Manishearth) [reply components/normalizer/src/lib.rs]: This is a good idea, I did one better, I wrote a seamless shim type `CanonicalCombiningClass(pub u8)` that gets imported in non-properties mode.

**Review by robertbastian — DISMISSED** [review]

**Review by sffc — DISMISSED** [review]

**Review by sffc — APPROVED** [review]

**Review by younies — APPROVED** [review]

**Review by sffc — APPROVED** [review]

**Pull request merged**